diffrax/brownian/tree.py

```diff
@@ -88,69 +144,173 @@ def __init__(
+            return (z * sqrt_len).astype(z.dtype)
+
+        def mult(z):
+            return z * (interval_len).astype(z.dtype)
+
+        return LevyVal(
+            h=jtu.tree_map(mult, x.h),
+            W=jtu.tree_map(sqrt_mult, x.W),
+            J=jtu.tree_map(mult, x.J),
+            H=jtu.tree_map(sqrt_mult, x.H),
+        )
+
     @eqx.filter_jit
     def evaluate(
-        self, t0: Scalar, t1: Optional[Scalar] = None, left: bool = True
-    ) -> PyTree[Array]:
-        del left
+        self,
+        t0: Scalar,
+        t1: Optional[Scalar] = None,
+        left: bool = True,
+        use_levy: bool = False,
+    ) -> LevyVal:
+        """Computes the Brownian increment between times `t0` and `t1`.
+
+        **Arguments:**
+            - `t0`: Start of interval
+
+            - `t1`: End of interval
+
+            - `left`: ignored since Brownian motion is continuous
+
+            - `use_levy`: If true, then the return type is a LevyVal encapsulating
+            the Brownian increment and the corresponding Lévy area if it was computed.
+        """
+
+        def _is_levy_val(obj):
+            return isinstance(obj, LevyVal)
+
         t0 = eqxi.nondifferentiable(t0, name="t0")
+        # map the interval [self.t0, self.t1] onto [0,1]
+        t0 = linear_rescale(self.t0, t0, self.t1)
+        levy_0 = self._evaluate(t0)
         if t1 is None:
-            return self._evaluate(t0)
+
+            if self.spacetime_levyarea:
+                levy_out = jtu.tree_map(_wj_to_wh_single, levy_0, is_leaf=_is_levy_val)
+            else:
+                levy_out = levy_0
+
         else:
             t1 = eqxi.nondifferentiable(t1, name="t1")
-            return jtu.tree_map(
-                lambda x, y: x - y,
-                self._evaluate(t1),
-                self._evaluate(t0),
+            # map the interval [self.t0, self.t1] onto [0,1]
+            t1 = linear_rescale(self.t0, t1, self.t1)
+            levy_1 = self._evaluate(t1)
+            levy_diff = (
+                _wj_to_wh_diff
+                if self.spacetime_levyarea
+                else lambda x, y: LevyVal(h=y.h - x.h, W=y.W - x.W)
             )
+            levy_out = jtu.tree_map(levy_diff, levy_0, levy_1, is_leaf=_is_levy_val)
 
-    def _evaluate(self, τ: Scalar) -> PyTree[Array]:
-        map_func = lambda key, shape: self._evaluate_leaf(key, τ, shape)
+        levy_out = _levy_tree_transpose(self.shape, self.spacetime_levyarea, levy_out)
+        # now map [0,1] back onto [self.t0, self.t1]
+        levy_out = self._denormalise_bm_inc(levy_out)
+        assert isinstance(levy_out, LevyVal)
+        return levy_out if use_levy else levy_out.W
+
+    def _evaluate(self, r: Scalar) -> PyTree[LevyVal]:
+        """Maps the _evaluate_leaf function at time τ using self.key onto self.shape"""
+        eqxi.error_if(
+            r,
+            r < 0,
+            "Cannot evaluate VirtualBrownianTree outside of its range [t0, t1].",
+        )
```

Comment:
Here you need to use the return value: `r = eqx.error_if(r, ...)`. The first argument is returned with the same value, but with the error "threaded on to it". (That is, the error condition is checked at some point between the input `r` being computed and the output `r` being used. And if the output is unused then this check is never performed.)

Reply:
Ohh, I see, thanks.